beater/middleware/timeout_middleware.go

```diff
@@ -0,0 +1,46 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package middleware
+
+import (
+	"context"
+
+	"github.com/pkg/errors"
+
+	"github.com/elastic/apm-server/beater/request"
+)
+
+// TimeoutMiddleware assumes that a context.Canceled error indicates a timed out
+// request. This could be caused by a either a client timeout or server timeout.
+// The middleware sets the Context.Result.
+func TimeoutMiddleware() Middleware {
+	tErr := errors.New("request timed out")
+	return func(h request.Handler) (request.Handler, error) {
+		return func(c *request.Context) {
+			h(c)
+
+			err := c.Request.Context().Err()
+			if errors.Is(err, context.Canceled) {
+				id := request.IDResponseErrorsTimeout
+				code := request.MapResultIDToStatus[id].Code
+
+				c.Result.Set(id, code, request.MapResultIDToStatus[id].Keyword, tErr.Error(), tErr)
```

Comment:
```suggestion
				c.Result.SetDefault(request.IDResponseErrorsTimeout)
```

SetDefault internally maps the ID to HTTP status code and message.